src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java

```diff
@@ -366,6 +369,92 @@ String readErrorMessage(InputStream in) throws IOException {
         return message.toString();
     }
 
+    /*
+     * Utility method to process the completion status after command execution.
+     * If we get IOE during previous command execution, delay throwing it until
+     * completion status have been read.
+     */
+    void processCompletionStatus(IOException ioe, String cmd, InputStream sis) throws AgentLoadException, IOException {
+        // Read the command completion status
+        int completionStatus;
+        try {
+            completionStatus = readInt(sis);
+        } catch (IOException x) {
+            sis.close();
+            if (ioe != null) {
+                throw ioe;
+            } else {
+                throw x;
+            }
+        }
+        if (completionStatus != 0) {
+            // read from the stream and use that as the error message
+            String message = readErrorMessage(sis);
+            sis.close();
+
+            // In the event of a protocol mismatch then the target VM
+            // returns a known error so that we can throw a reasonable
+            // error.
+            if (completionStatus == ATTACH_ERROR_BADVERSION) {
+                throw new IOException("Protocol mismatch with target VM");
+            }
+
+            // Special-case the "load" command so that the right exception is
+            // thrown.
+            if (cmd.equals("load")) {
+                String msg = "Failed to load agent library";
+                if (!message.isEmpty()) {
+                    msg += ": " + message;
+                }
+                throw new AgentLoadException(msg);
+            } else {
+                if (message.isEmpty()) {
+                    message = "Command failed in target VM";
+                }
+                throw new AttachOperationFailedException(message);
+            }
+        }
+    }
+
+    /*
+     * InputStream for the socket connection to get target VM
+     */
+    static abstract class SocketInputStream extends InputStream {
```

Comment:
Use blessed modifiers order
```suggestion
    abstract static class SocketInputStream extends InputStream {
```